packages/expo-network/android/src/main/java/expo/modules/network/NetworkModule.kt

```diff
@@ -21,50 +22,41 @@ import java.nio.ByteOrder
 
 private val TAG = NetworkModule::class.java.simpleName
 
+internal const val NETWORK_STATE_EVENT_NAME = "onNetworkStateChanged"
+
 class NetworkModule : Module() {
   private val context: Context
     get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()
 
   override fun definition() = ModuleDefinition {
     Name("ExpoNetwork")
 
-    AsyncFunction("getNetworkStateAsync") { promise: Promise ->
-      val result = Bundle()
-      val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
-
-      try {
-        if (Build.VERSION.SDK_INT < 29) { // use getActiveNetworkInfo before api level 29
-          val netInfo = connectivityManager.activeNetworkInfo
-          val connectionType = getConnectionType(netInfo)
+    Events(NETWORK_STATE_EVENT_NAME)
 
-          result.apply {
-            putBoolean("isInternetReachable", netInfo!!.isConnected)
-            putString("type", connectionType.value)
-            putBoolean("isConnected", connectionType.isDefined)
-          }
-
-          promise.resolve(result)
-        } else {
-          val network = connectivityManager.activeNetwork
-          val isInternetReachable = network != null
-
-          val connectionType = if (isInternetReachable) {
-            val netCapabilities = connectivityManager.getNetworkCapabilities(network)
-            getConnectionType(netCapabilities)
-          } else {
-            null
+    OnCreate {
+      val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+      val networkRequest = NetworkRequest.Builder().build()
+      connectivityManager.registerNetworkCallback(
+        networkRequest,
+        object : ConnectivityManager.NetworkCallback() {
+          override fun onAvailable(network: android.net.Network) {
+            emitNetworkState()
           }
 
-          result.apply {
-            putString("type", connectionType?.value ?: NetworkStateType.NONE.value)
-            putBoolean("isInternetReachable", isInternetReachable)
-            putBoolean("isConnected", connectionType != null && connectionType.isDefined)
+          override fun onLost(network: android.net.Network) {
+            emitNetworkState()
           }
-          promise.resolve(result)
         }
-      } catch (e: Exception) {
-        throw NetworkAccessException(e)
-      }
+      )
+    }
+
+    OnDestroy {
+      val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+      connectivityManager.unregisterNetworkCallback(ConnectivityManager.NetworkCallback())
+    }
+
+    AsyncFunction("getNetworkStateAsync") { promise: Promise ->
+      promise.resolve(fetchNetworkState())
```

Comment:
```suggestion
    AsyncFunction("getNetworkStateAsync") { 
      return@AsyncFunction fetchNetworkState()
```
We can drop the promise and return this directly